### ISSUE TITLE
USB-OTG support gone?

### ISSUE THREAD

**Issue by dcmcshan** [lvgl-micropython/lvgl_micropython#231]
**Describe the bug**
--usb-otg support used to work, now does not?

**Expected behavior**
After loading firmware and a reset, I expect to see a new USB port.  Build flashes and verifies correctly.  Loading a previous image does the right thing.

**Exact make and model number of the MCU that you are compiling for or the firmware is running on. **

* Make: ESP32S3
* Model: -N8R8

**For ESP32 MCU's The PSRAM and FLASH SPI type, quad SPI or octal SPI.**

* PSRAM: octal 8MB
* FLASH:  quad 8MB

I need to know the OS and OS version of the machine that compiled the binary. If using a VM then I need to know the OS and OS version the VM is running. (WSL == VM)
 - OS: MacOS

**Build Command**
python make.py esp32 BOARD=ESP32_GENERIC_S3 BOARD_VARIANT=SPIRAM_OCT --flash-size=8 DISPLAY=ST7735 --usb-otg --dual-core-threads

**Comment by dcmcshan** [comment]
I think something in 1.24 breaks USB-OTG, as I can load stock 1.23 and it works, but 1.24.1 does not.

**Comment by kdschlosser** [comment]
I know there was some changes to micropython that allow for better handling of the USB-OTG so I will have to make adjustments to get it running again. Give me a day or 2 in order to get that done.

**Comment by kdschlosser** [comment]
I saw that you opened up an issue with the folks over at MicroPython. I have tried on more than one occasion to get them to reconsider how they are handling the USB-CDC as well as the UART. The entire way it is currently set up makes assumptions as to what is to be used and it also locks things in place and they cannot be changed by the user without making some serious modifications to MicroPython. They have not listened to me about this and the changes they have made do not fix anything. They still make the assumption that if the MCU supports USB-CDC then that is what is to be used for the REPL and it can ONLY be used for that singular purpose. It's a bad design what they have done and they are not willing to change it at all as it would appear. The recent changes they made solidify that as it has made it even harder to undo.

**Comment by dcmcshan** [comment]
Yeah, I'm a bit concerned that they are proceeding on a path that breaks me.  Do you think you can help me get this up and running?  I can get you a platform to play with....

**Comment by kdschlosser** [comment]
yeah it's not a big deal to fix. It would be easier if I had whatever it is you are working on to test with to make sure it is 100% functional...

I have not spent any time looking over the code changes that have been made to MicroPython with respect to the CDC stuff. I did a quick look and it looks like it is going to take a little bit more effort to solve the issue than it did previously because of the changes that were made.

What I want to do is expose the API to be able to use the OTG for whatever it is you want. This would include directing the REPL to a CDC connection or directing the output to the UART...

Let me give you a tidbit of background on the OTG for the ESP32 coupled with the UART.

Before the OTG was added UART0 is what was used to flash the firmware. this required a TTL to USB bridge IC to be able to do that. The MCU would need to be put into programming mode. Once the MCU had it's firmware flashed there is no requirement to keep a TTL to USB bridge IC connected to the UART0 pins so the pins could be used for some other purpose once the firmware was flashed.

The OTG functions in the same exact manner. Built into the ESP32 is a small USB CDC stack that operates for the sole purpose of flashing firmware. Once the firmware is flashed there is no requirement to keep on using those pins for the USB. What that means is there should also be no requirement locking the use of those pins to either a UART or the CDC, There isn't even a requirement to use those pins for flashing because it is able to do that over a WiFi connection. MicroPython also has the webrepl tho I believe that bounces off one of their servers which is not something that really appeals to me. It would not be too hard to write a point to point telnet connection for the repl. My point being is there should be no requirement to define some kind of output for the REPL at compile time. It is something that should be able to be done at runtime using the boot.py script since after all that is why that script exists in MicroPython.

The issue with the current design is the requirement to have some kind of a hardware connection for the REPL even when using the webrepl the hardware connection MUST still exist even if it goes unused.

The ability to bake python files into the firmware solidifies that statement because the boot.py file is able to be written and baked into the firmware so it gets uploaded when the firmware gets uploaded so there is no need to transfer the boot.py file using the typical way of sending files using the REPL.

Best way for them to handle it is to write boot.py template files, one for UART, one for OTG CDC and the 3rd for the webrepl. Those templates get added to a boot.py file that is created at compile time and what templates get added is based on board definitions or what the user has turned on/off manually using command line switches. Optionally a user could also provide their own boot.py file if they should so choose to.

That's how I see a proper system working. In all reality because MCU's like the ESP32 having NVS memory settings are able to be saved that can be loaded at boot time that could dictate how to use a set of pins. You could have some kind of a connector that is normally used for expansion purposes but via a user setting maybe by pressing and holding 2 or 3 buttons at the same time would change that port over to be able to be used to access the REPL using a TTL to USB bridge module or some kind of a plugin expansion board that provides the ability to plug in directly to USB so you can connect using CDC...

**Comment by kdschlosser** [comment]
The USB OTG on the ESP32 is only limited by the software that is running on the ESP32. The PORT can be used as a host so you can plug in Keyboards and Mice into the ESP32 and it can also act as a device. It can emulate a keyboard or mouse. Now here is a real head splitter. It is able to be a host and a device at the same time..... You may ask why would you want to do that. Well if you want to plug a Keyboard into the ESP you can do that, But then if you needed to connect to the REPL for some reason you could unplug the keyboard and then plug the the ESP32 into a PC and now you can access the REPL. The pins are not limited to only being able to do a single thing. MicroPython however is not allowing it to be used in this manner.

**Comment by dcmcshan** [comment]
I like that flexibility to us that usb otg in various ways and can easily imagining adding accessories to the device over usb.

Fo you think we'll be able to configure usb otg in boot.py in spite of what micropython is doing?

I will g eat you some hardware soon.

**Comment by kdschlosser** [comment]
OK I updated the build script so hopefully it should now work with the USB-CDC..

If you give me the build command you normally use I will write something up that I think will be very beneficial for you to use.

I did change the command from `--usb-otg` to `--enable-cdc-repl=y/n`. There is also `--enable-uart-repl=y/n` as well. Yes you can have both enabled at the same time. If you do not specifically set `--enable-uart-repl=n` then it falls back to whatever the board is preset to in MicroPython. For the "GENERIC" boards it will be turned on.

**Comment by dcmcshan** [comment]
python3 make.py esp32 clean  BOARD=ESP32_GENERIC_S3 BOARD_VARIANT=SPIRAM_OCT --flash-size=8 DISPLAY=ST7735 --usb-otg

is what I've been using

**Comment by dcmcshan** [comment]
Getting ESP-IDF build Environment
/Applications/Xcode.app/Contents/Developer/usr/bin/make: unrecognized option `--enable-cdc-repl=y' 

Did you check in?

**Comment by kdschlosser** [comment]
It should be working now. I had a couple of things I needed to fix.

**Comment by dcmcshan** [comment]
That does seem to work.  Of interest, perhaps, is that it used to be that DFU connected as /dev/tty.usbmodem2101 and repl would live on /dev/tty.usbmodem1234561.  The actual values are somewhere arbitrary, but it was nice that they changed for scripting purposes.  I know there is a way to set the USB host name... any idea where that might live?

**Comment by kdschlosser** [comment]
you need to slow yourself.. you are getting ahead of me here...

OK so I added a nifty little thing just for you.

You have the ability to create a custom board in a directory that is not in the repository path. This directory is able to be passed to the build script.

in that directory you are going to need to have several files. These files are what define what the board is. One of the things you are going to be able to do is to override a grip of the default MicroPython settings. Of those things are the VID, PID, manufacturer name and device name that appear for the USB.. YAY!!!

By supplying a custom board you will only have to pass the path to the custom board to the build script. No other options will need to be passed because those options you are going to be able to set yourself in those board files.

The build command will look like this..

```
python3 make.py esp32 --custom-board-path=/absolute/path/to/CUSTOM BOARD
```

All of your settings like the flash size, flash speed, spiram speed, spiram type  and any other ESP-IDF config setting you want to use will be in one of those files.

MicroPython config settings will be in another file. You will be able to override things like the default UART speed (if using the UART repl) what kind of repl to use (cdc, jtag or uart), usb pid, usb vid, usb identification strings, etc....

There is going to be a file where you define the partitions how you want.

The same things you see in the current board specification files.

You however do not have the option to specify a board variant. That you will do by creating a different custom board.

But to answer your direct question. You want to add the following things (if you want to change them) to the top of the `lib/micropython/ports/esp32/mpconfigport.h` file

```c
#ifdef MICROPY_HW_USB_VID
#undef MICROPY_HW_USB_VID
#endif
#define MICROPY_HW_USB_VID    ({custom vid})
```
replacing `{custom vid}` with the hex for your custom VID

```c
#ifdef MICROPY_HW_USB_PID
#undef MICROPY_HW_USB_PID
#endif
#define MICROPY_HW_USB_PID    ({custom pid})
```
replacing `{custom pid}` with the hex for your custom PID

```c
#ifdef MICROPY_HW_USB_MANUFACTURER_STRING
#undef MICROPY_HW_USB_MANUFACTURER_STRING
#endif
#define MICROPY_HW_USB_MANUFACTURER_STRING  "{custom manufacturer name}"
```
replacing `{custom manufacturer name}` with the name you want to use

```c
#ifdef MICROPY_HW_USB_PRODUCT_FS_STRING
#undef MICROPY_HW_USB_PRODUCT_FS_STRING
#endif
#define MICROPY_HW_USB_PRODUCT_FS_STRING    "{custom product name}"
```
replacing `{custom product name}` with the name you want to use
